src/drivers/driver.ts

```diff
@@ -614,10 +615,64 @@ export abstract class CMakeDriver implements vscode.Disposable {
     await this.doPreCleanConfigure();
     this.configRunning = false;
 
-    return this.configure(extra_args, consumer);
+    return this.configure(trigger, extra_args, consumer);
   }
 
-  async configure(extra_args: string[], consumer?: proc.OutputConsumer, withoutCmakeSettings:boolean = false): Promise<number> {
+  async testCompilerVersion(program: string, cwd: string, arg: string | undefined, regexp: RegExp): Promise<string> {
+    const args = [];
+    if (arg) {
+      args.push(arg);
+    }
+    const child = this.executeCommand(program, args, undefined, {silent: true, cwd});
+    try {
+      const result = await child.result;
+      log.debug(localize('command.version.test.return.code', 'Command version test return code {0}', nullableValueToString(result.retc)));
+      // cl.exe version information is printed on stderr when no source files are given
+      const versionLine = (program === "cl") ? result.stderr : result.stdout; //.split("\\n")[0];
+      const match = regexp.exec(versionLine);
+      return match ? match[1] : "error";
+    } catch (e) {
+      const e2: NodeJS.ErrnoException = e;
+      log.debug(localize('compiler.version.return.code', 'Compiler version test return code {0}', nullableValueToString(e2.code)));
+      if (e2.code === 'error') {
+        return e2.code;
+      }
+      throw e;
+    }
+  }
+
+  async getCompilerVersion(compilerPath: string) : Promise<string | undefined> {
+    const compilerName = path.parse(compilerPath).name;
+    const compilerDir = path.parse(compilerPath).dir;
+    let versionSwitch;
+    let versionRegexp;
+    switch (compilerName) {
```

Comment:
I don't think a `switch` will work very well for us because there are tons of variants of gcc compilers out there and we will miss a lot of them if we only take the ones with these exact names.  Take a look at kit.ts which does some version parsing.  Maybe there's a way to reuse some of the code.  In the cpptools closed-source code, there is some logic you should look at to get these extra compiler names.

Reply:
Ok, I got the list of compiler names from cpptools (src\vc\projbld\CMake\Package\Telemetry\Telemetry.cs) and I hardcoded each version switch and regexp scenario based on man commands and internet searches (I wasn't able to install all of them to really see what works).

Reply:
This doesn't cover cross compilers with triplet names (and I don't think a `switch` is the right tool for this since there are a lot of combinations)
For example: `arm-none-eabi-gcc`

---
In reply to: [517393239](https://github.com/Microsoft/vscode-cmake-tools/pull/1503#discussion_r517393239) [](ancestors = 517393239)

Reply:
Maybe this logic can be added to the `default` case.  Look at the code for cpptools for how we report this compilers.  Search for the variable `compiler_allowed_list` in the code.

---
In reply to: [517487911](https://github.com/Microsoft/vscode-cmake-tools/pull/1503#discussion_r517487911) [](ancestors = 517487911,517393239)

Reply:
I didn't find any compiler_allowed_list, but SupportedDefaultCompilers. Did you mean that one?

Reply:
Are you looking in the C++ code, or the TypeScript?  It's not in the TypeScript.